src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -220,19 +276,27 @@ ObjectDataHandler::write_ret do_insertions(
 	DEBUGT("map existing extent: laddr {} len {} {}",
 	       ctx.t, region.addr, region.len, *region.existing_paddr);
 	return ctx.tm.map_existing_extent<ObjectDataBlock>(
-	  ctx.t, region.addr, *region.existing_paddr, region.len
+	  ctx.t, region.addr, *region.existing_paddr,
+	  region.len, {region.indirect_key, std::move(region.existing_pin)}
```

Comment:
> I think refactoring TM::map_existing_extent() may not be enough to avoid exposing physical addresses to ObjectDataHandler, because there are places in ObjectDataHandler where we need to distinguish the pins by their "zeroness". Maybe the whole ObjectDataHandler's way of handling lba mappings should be reconsidered.

I may not fully follow the issue here. IIUC `P_ADDR_ZERO` isn't real physical address so maybe we can leave it along for now. I'm not sure if there are any gaps in hiding indirect laddr from ObjectDataHandler based on `TransactionManager::remap_pin()` and `overwrite_pin()`. Please let me know so we can discuss as early as possible.

> Actually, @zhscn and I have been considering some kind of TransactionManager::transform(lba mapping set) interface, which enables ObjectDataHandler to pass the lba mapping set that they want to overwrite and do all the lba mapping transformations altogether. If this is feasible, we should have an interface that's both semantically clear and efficient.

I'm not sure why `TransactionManager::remap_pin()` and `overwrite_pin()` are not sufficient here, or just barely sufficient? Can you elaborate with real examples?

> Yeah, of course:-)

Please also help review https://github.com/ceph/ceph/pull/51179 and let me know if there is anything unreasonable from cloning perspective.

In summary, I think it is important to agree on the interface at TM level, or at least agree on a reasonable-enough version of interface for now. So we won't further complicate implementations based on apprantly inappropriate designs, then back to fix it, or going on with complete different directions.

Reply:
> > I think refactoring TM::map_existing_extent() may not be enough to avoid exposing physical addresses to ObjectDataHandler, because there are places in ObjectDataHandler where we need to distinguish the pins by their "zeroness". Maybe the whole ObjectDataHandler's way of handling lba mappings should be reconsidered.
> 
> I may not fully follow the issue here. IIUC `P_ADDR_ZERO` isn't real physical address so maybe we can leave it along for now. And real physical address is only used with `TM::map_existing_extent()` which will be addressed by #51179. I'm not sure if there are any gaps in hiding indirect laddr from ObjectDataHandler based on `TransactionManager::remap_pin()` and `overwrite_pin()`. Please let me know so we can discuss as early as possible.
> 
> > Actually, @zhscn and I have been considering some kind of TransactionManager::transform(lba mapping set) interface, which enables ObjectDataHandler to pass the lba mapping set that they want to overwrite and do all the lba mapping transformations altogether. If this is feasible, we should have an interface that's both semantically clear and efficient.
> 
> I'm not sure why `TransactionManager::remap_pin()` and `overwrite_pin()` are not sufficient here, or just barely sufficient? Can you elaborate with real examples?

I just think, if we really want to hide physical addresses from anything above TM, maybe we should allow anything above TM to get the paddr of an lba mapping, whether it is P_ADDR_ZERO or not. What do you think?

> 
> > Yeah, of course:-)
> 
> Please also help review #51179 and let me know if there is anything unreasonable from cloning perspective.

Ok:-)
> 
> In summary, I think it is important to agree on the interface at TM level, or at least agree on a reasonable-enough version of interface for now. So we won't further complicate implementations based on apparantly inappropriate designs, then back to fix it, or going on with complete incompatible directions.

Yeah, totally agree.

Reply:
> I just think, if we really want to hide physical addresses from anything above TM, maybe we shouldn't allow anything above TM to get the paddr of an lba mapping, whether it is P_ADDR_ZERO or not. What do you think?

Hmm, seems to me it is not a critical issue. Introducing a `pin.is_zero()` should be sufficient, right? Completely enforce that users above TM cannot get physical-address will need to adjust the design of interfaces.

Reply:
> > I just think, if we really want to hide physical addresses from anything above TM, maybe we shouldn't allow anything above TM to get the paddr of an lba mapping, whether it is P_ADDR_ZERO or not. What do you think?
> 
> Hmm, seems to me it is not a critical issue. Introducing a `pin.is_zero()` should be sufficient, right? Completely enforce that users above TM cannot get physical-address will need to adjust the design of interfaces.

Yeah, you're right.

Reply:
@cyx1231st There's another reason for `TM::transform()`, but it's performance related one. Currently, ObjectDataHandler's modification generally goes the following way:
1. get lba mappings that are related to the modifications (1 lba tree search);
2. do all the lba mapping transformation calculations
3. do removals(1 lba tree search for each removal)
4. do insertions(1 lba tree search for each insertion)

If we can implement the `TM::transform()` interface, all the above processes can be done with only 1 lba tree search

Reply:
It can be experimented independently. And delta-based data overwrite https://github.com/ceph/ceph/pull/51179#discussion_r1209665126 seems better at minimizing the lba overhead.

Reply:
I think the delta-based approach works for modifying a part of the existing extent. However, if an overwrite spans multiple existing extents, there will still be multiple lba searches.

Reply:
Sounds good, probably need to understand how to make it compatible with delta-based overwrite at the level of ObjectDataBlock in ObjectDataHandler.

Anyway, it seems simpler to me to start with a plain pin-based overwrite and remap interfaces at TM level for both lazy-read and the initial laddr indirection support.

Reply:
> Anyway, it seems simpler to me to start with a plain pin-based overwrite and remap interfaces at TM level for both lazy-read and the initial laddr indirection support.

Yeah, that's for sure.